api/src/test/java/org/openmrs/validator/ObsValidatorTest.java

```diff
@@ -513,4 +517,23 @@ public void validate_shouldPassValidationIfValueTextIsLessThanTheMaximumLength()
 		Assert.assertFalse(errors.hasFieldErrors("obsDatetime"));
 		Assert.assertFalse(errors.hasFieldErrors("valueText"));
 	}
+
+	@Test
+	public void validate_shouldFailValidationForDifferentObsPersonAndWithEncounterPatient() {
+		EncounterService es = Context.getEncounterService();
+		PersonService ps = Context.getPersonService();
+
+		Obs obs = new Obs(1);
+		obs.setPerson(ps.getPerson(2)); // Setting this person (2) as different from that of encounter person
+		obs.setEncounter(es.getEncounter( 3 ));
+		obs.getEncounter().setPatient(new Patient( 3 ) ); //Setting encounter person(3) as different from that of obs person
+
+		assertFalse( Objects.equals(obs.getPerson(),obs.getEncounter().getPatient()) ); //checking whether the two person objects are different
+		
+		Errors errors = new BindException(obs, "obs");
+		new ObsValidator().validate(obs, errors);
+
+		assertTrue(errors.hasFieldErrors("person"));
```

Comment:
use another assertion to ensure this test will fail if the field does not have the exact error you expect. imagine the person field would still have an error but due to another error code. the test would still pass.

`assertThat(errors, hasFieldErrors("person", "error.obs.encounter.mismatch"));`
 
for details see
https://github.com/openmrs/openmrs-core/blob/master/api/src/test/java/org/openmrs/validator/VisitValidatorTest.java#L545